File: engine/src/Unit.ts
// 运行时单位实例
// Unity 迁移：改成 MonoBehaviour，pos 换 Transform.position

import { Vec2 } from "./Vec2.ts";
import type { Side } from "./Enums.ts";
import type { UnitTemplate } from "./Templates.ts";
import { DamageCalc } from "./DamageCalc.ts";
import type { Building } from "./Building.ts";

export type AttackTarget = Unit | Building;

export class Unit {
  static _nextId = 1;

  readonly id: number;
  readonly tpl: UnitTemplate;
  readonly side: Side;
  pos: Vec2;

  hp: number;
  mana: number;

  /** 下一次可出手的绝对时间 (秒) */
  nextAttackReadyAt = 0;
  /** 攻击动画关键帧触发时间；null 表示当前不在挥击 */
  swingAt: number | null = null;
  /** 正在挥向的目标（Unit 或 Building） */
  swingTarget: AttackTarget | null = null;

  alive = true;
  deadAt = 0;

  constructor(tpl: UnitTemplate, side: Side, pos: Vec2) {
    this.id = Unit._nextId++;
    this.tpl = tpl;
    this.side = side;
    this.pos = pos;
    this.hp = tpl.hp;
    this.mana = tpl.manaMax ?? 0;
  }

  takeDamage(raw: number, atkType: import("./Enums.ts").DamageType): number {
    const actual = DamageCalc.calc(raw, atkType, this.tpl.defense.armor, this.tpl.defense.armorType);
    this.hp -= actual;
    if (this.hp <= 0) {
      this.hp = 0;
      this.alive = false;
    }
    return actual;
  }

  distTo(other: { pos: Vec2 }): number {
    return Vec2.distance(this.pos, other.pos);
  }

  /** 前进方向 (+y 为 Left 攻向 Right) */
  forwardSign(): number { return this.side === 0 ? 1 : -1; }
}

File: unity/Assets/Mg01/Sim/Unit.cs
// Unity 端 Unit 运行时实例。MonoBehaviour 化：transform.position 即 pos
// 与 engine/src/Unit.ts 一一对应

using System.Collections.Generic;
using UnityEngine;
using Mg01.Core;
using Mg01.Templates;

namespace Mg01.Sim
{
    public class Unit : MonoBehaviour
    {
        public static int _nextId = 1;

        public int id;
        public UnitTemplate tpl;
        public Side side;

        public float hp;
        public float mana;

        public float nextAttackReadyAt;
        public float? swingAt;            // null = 不在挥击
        public Object swingTarget;        // Unit 或 Building (轻量起见用 Object)

        public List<Buff> buffs = new();
        public float nextSpellReadyAt;
        public PendingSpell pendingSpell; // null = 没在施法

        public bool alive = true;
        public float deadAt;

        /// <summary>构造期：直接 new (Editor)；运行时请用 BattleSim.SpawnUnit Instantiate</summary>
        public void Init(UnitTemplate t, Side s, Vector2 pos)
        {
            id = _nextId++;
            tpl = t;
            side = s;
            transform.position = pos;
            hp = t.hp;
            mana = t.manaMax;
        }

        /// <summary>对应 Unit.takeDamage</summary>
        public float TakeDamage(float raw, DamageType atk)
        {
            float armor = tpl.defense.armor;
            foreach (var b in buffs) armor += b.armorFlat;
            var armorE = EnumNames.ParseArmorType(tpl.defense.armorType);
            float actual = DamageCalc.Calc(raw, atk, armor, armorE);
            hp -= actual;
            if (hp <= 0f) { hp = 0f; alive = false; }
            return actual;
        }

        /// <summary>+y = Left 攻向 Right</summary>
        public int ForwardSign() => side == Side.Left ? 1 : -1;

        public Vector2 Pos => transform.position;
    }

    public class PendingSpell
    {
        public SpellTemplate sp;
        public Unit target;
        public float castAt;
    }
}
